File: Python/_1035_Uncrossed_Lines/Solution.py
# coding=utf-8
from typing import List


# 执行用时：200 ms, 在所有 Python3 提交中击败了64.75%的用户
# 内存消耗：15.2 MB, 在所有 Python3 提交中击败了47.12%的用户
class Solution:
    def maxUncrossedLines(self, nums1: List[int], nums2: List[int]) -> int:
        n = len(nums1); m = len(nums2)
        f = [[0 for _ in range(m + 1)] for _ in range(n + 1)]
        for i in range(1, n + 1):
            for j in range(1, m + 1):
                f[i][j] = max(f[i - 1][j], f[i][j - 1])
                if nums1[i - 1] == nums2[j - 1]:
                    f[i][j] = max(f[i][j], f[i - 1][j - 1] + 1)
        return f[n][m]


if __name__ == "__main__":
    print(Solution().maxUncrossedLines([1, 4, 2], [1, 2, 4]))  # 2

File: Java/src/_1035_Uncrossed_Lines/Solution.java
package _1035_Uncrossed_Lines;

/**
 * Date: 2021/5/21 8:40
 * 执行用时：6 ms, 在所有 Java 提交中击败了60.90%的用户
 * 内存消耗：37.8 MB, 在所有 Java 提交中击败了72.30%的用户
 */
public class Solution {

    public int maxUncrossedLines(int[] nums1, int[] nums2) {
        int n = nums1.length, m = nums2.length;
        int[][] f = new int[n + 1][m + 1];
        for (int i = 1; i <= n; i++)
            for (int j = 1; j <= m; j++) {
                f[i][j] = Math.max(f[i - 1][j], f[i][j - 1]);
                if (nums1[i - 1] == nums2[j - 1]) f[i][j] = Math.max(f[i][j], f[i - 1][j - 1] + 1);
            }
        return f[n][m];
    }

    public static void main(String[] args) {

        int[] nums1 = {1, 4, 2}, nums2 = {1, 2, 4};
        System.out.println((new Solution()).maxUncrossedLines(nums1, nums2));  // 2
    }
}
